File: src/lib/downloader/thread-ids.ts
/** XenForo / forum URL helpers: /forums/name.12/  /threads/title.345/  /posts/678  #post-678 */

export interface ThreadIds {
  forumId: string | null;
  threadId: string | null;
  postId: string | null;
}

export function parseThreadIds(url: string): ThreadIds {
  try {
    const u = new URL(url);
    const path = u.pathname;
    const forum = path.match(/\/forums\/(?:[^/]*?\.)?(\d+)/i);
    const thread = path.match(/\/threads\/(?:[^/]*?\.)?(\d+)/i);
    const postPath = path.match(/\/posts\/(\d+)/i);
    const postHash = u.hash.match(/post-(\d+)/i);
    const postInThread = path.match(/\/threads\/[^/]+\/(?:page-\d+\/)?post-(\d+)/i);
    return {
      forumId: forum?.[1] ?? u.searchParams.get("forum_id") ?? u.searchParams.get("f"),
      threadId: thread?.[1] ?? u.searchParams.get("thread_id") ?? u.searchParams.get("t"),
      postId:
        postPath?.[1] ??
        postInThread?.[1] ??
        postHash?.[1] ??
        u.searchParams.get("post_id") ??
        u.searchParams.get("p"),
    };
  } catch {
    return { forumId: null, threadId: null, postId: null };
  }
}

export function postIdFromAttrs(attrs: string): string | null {
  const id = attrs.match(/\b(?:id|data-content|data-post-id|data-lb-id)=["'](?:js-)?post-?(\d+)/i);
  if (id?.[1]) return id[1];
  const data = attrs.match(/\bdata-post=["'](\d+)/i);
  return data?.[1] ?? null;
}

export function forumIdFromHtml(html: string): string | null {
  const crumb = html.match(/\/forums\/(?:[^"'/]*?\.)?(\d+)/i);
  return crumb?.[1] ?? null;
}

File: desktop/src/SimpDownloader.Core/ThreadIds.cs
using System.Text.RegularExpressions;

namespace SimpDownloader.Core;

public readonly record struct ThreadIds(string? ForumId, string? ThreadId, string? PostId);

public static class ThreadIdParser
{
    public static ThreadIds Parse(string url)
    {
        if (!Uri.TryCreate(url, UriKind.Absolute, out var u)) return default;
        var path = u.AbsolutePath;
        string? Forum() => M(path, @"/forums/(?:[^/]*?\.)?(\d+)");
        string? Thread() => M(path, @"/threads/(?:[^/]*?\.)?(\d+)");
        string? Post() => M(path, @"/posts/(\d+)") ?? M(u.Fragment, @"post-(\d+)");
        return new ThreadIds(Forum() ?? Q(u, "forum_id") ?? Q(u, "f"), Thread() ?? Q(u, "thread_id") ?? Q(u, "t"), Post() ?? Q(u, "post_id") ?? Q(u, "p"));
    }

    public static string? PostFromAttrs(string attrs)
    {
        var m = Regex.Match(attrs, @"\b(?:id|data-content|data-post-id)=[""'](?:js-)?post-?(\d+)", RegexOptions.IgnoreCase);
        return m.Success ? m.Groups[1].Value : null;
    }

    static string? M(string input, string pattern)
    {
        var m = Regex.Match(input, pattern, RegexOptions.IgnoreCase);
        return m.Success && m.Groups[1].Value.Length > 0 ? m.Groups[1].Value : null;
    }

    static string? Q(Uri u, string key)
    {
        foreach (var part in u.Query.TrimStart('?').Split('&', StringSplitOptions.RemoveEmptyEntries))
        {
            var kv = part.Split('=', 2);
            if (kv.Length == 2 && kv[0].Equals(key, StringComparison.OrdinalIgnoreCase))
                return Uri.UnescapeDataString(kv[1]);
        }
        return null;
    }
}
